src/Illuminate/Database/Eloquent/Concerns/PreventsCircularRecursion.php

```diff
@@ -0,0 +1,74 @@
+<?php
+
+namespace Illuminate\Database\Eloquent\Concerns;
+
+use Illuminate\Support\Arr;
+use Illuminate\Support\Onceable;
+
+trait PreventsCircularRecursion
+{
+    /**
+     * The cache of objects processed to prevent infinite recursion.
+     *
+     * @var \WeakMap<static, array<string, mixed>>
+     */
+    protected static $recursionCache;
+
+    /**
+     * Get the current recursion cache being used by the model.
+     *
+     * @return \WeakMap
+     */
+    protected static function getRecursionCache()
+    {
+        return static::$recursionCache ??= new \WeakMap();
+    }
+
+    /**
+     * Get the current stack of methods being called recursively.
+     *
+     * @param  object  $object
+     * @return array
+     */
+    protected static function getRecursiveCallStack($object): array
+    {
+        return static::getRecursionCache()->offsetExists($object)
+            ? static::getRecursionCache()->offsetGet($object)
+            : [];
+    }
+
+    /**
+     * Prevent a method from being called multiple times on the same object within the same call stack.
+     *
+     * @param  callable  $callback
+     * @param  mixed  $default
+     * @return mixed
+     */
+    protected function once($callback, $default = null)
```

Comment:
I do think following laravel's pattern this should be renamed to `withoutInfiniteRecursion` or the like

As you've explained it doesn't really have much to do with the other once helper and could be confusing

Reply:
I had actuall named it `withoutRecursion()` originally then changed it to `once()`

Reply:
That also works, as long as it makes the code more readable in `Model` so you can understand without having to take a look inside the function

Reply:
I've updated this back to `withoutRecursion()`